### PR TITLE
mpt: utf-8 support, perplexity testing, repeat penalty sampling

### PR DESCRIPTION
**Pull request by klosax**
Follow up to pr #179 .  Fixes issues #170 , #55 . 

A better solution that will not affect the other examples that uses gpt_tokenize

- Revert encoding of input to utf-8 in gpt_tokenize  
- Add decoding of utf-8 tokens on load in mpt main.
- Small change to mpt import script
- Add perplexity testing
- Add repeat penalty sampling

### CODE REVIEW

File: examples/common.cpp [hunk 1]
```diff
@@ -212,38 +212,40 @@ void gpt_vocab::add_special_token(const std::string & token) {
...
+			out.push_back( ((u8[i] & 0xf) << 12) | ((u8[i+1] & 0x3f) << 6) | (u8[i+2] & 0x3f) );
+		} else if (tmp < 0xf8UL) {
+			elem_len = 4;
+			out.push_back( ((u8[i] & 0x7) << 18) | ((u8[i+1] & 0x3f) << 12)	| ((u8[i+2] & 0x3f) << 6) | (u8[i+3] & 0x3f) );
+		} else {
+		
+            		printf("Invalid Unicode code point\n");
+            		break;
+		}
+	}
```
Comment (ggerganov) [inline examples/common.cpp:240]: Fix the indentation in this function

Comment (klosax) [reply examples/common.cpp]: good?

**Comment by zenixls2** [comment]
This matches the implementation in Transformer and should be back-ported to llama.cpp's examples/main and common.h/cpp.

File: examples/mpt/main.cpp [hunk 2]
```diff
@@ -64,6 +64,105 @@ struct mpt_model {
     std::map<std::string, struct ggml_tensor *> tensors;
 };
 
+struct mpt_params {
```
Comment (zenixls2) [inline examples/mpt/main.cpp:67]: since this is in cpp, is it possible to extend the `struct gpt_params`?

File: examples/common.cpp [hunk 3]
```diff
@@ -407,6 +408,129 @@ gpt_vocab::id gpt_sample_top_k_top_p(
...
+            if (plogits[i] > max_logit) {
+                max_logit = plogits[i];
+                max_id = i;
+            }
+        }
+        return max_id;
+    }
+
+
+    std::vector<std::pair<double, gpt_vocab::id>> logits_id;
```
Comment (zenixls2) [inline examples/common.cpp:445]: logit is float. Maybe use `std::pair<float, gpt_vocab::id>`?

**Comment by zenixls2** [comment]
also in llama.cpp/examples/main/main.cpp line 423
and in llama.cpp/llama.cpp line 1833

implements similar approach that makes use of the `sorted` flag to speed-up.

File: examples/mpt/perplexity.cpp [hunk 4]
```diff
@@ -0,0 +1,809 @@
...
+        printf("%s:     load time = %8.2f ms\n", __func__, t_load_us / 1000.0f);
+        printf("%s:  eval time = %8.2f ms / %.2f ms per token\n", __func__, t_predict_us / 1000.0f,
+               t_predict_us / 1000.0f / (n_chunk * params.n_ctx) );
+        printf("%s:    total time = %8.2f ms\n", __func__, (t_main_end_us - t_main_start_us) / 1000.0f);
+    }
+
+    ggml_free(model.ctx);
+
+    return 0;
+}
```
Comment (ggerganov) [inline examples/mpt/perplexity.cpp:809]: No need to duplicate `main.cpp` in `perplexity.cpp`.
Just add a `--perplexity` CLI argument to `main.cpp`

**Review by ggerganov — CHANGES_REQUESTED** [review]

File: examples/mpt/perplexity.cpp (hunk 4 above)
Comment (klosax) [reply examples/mpt/perplexity.cpp]: Done.

**Review by ggerganov — APPROVED** [review]

**Pull request merged**